File: quicksort.cpp
#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>
#include <time.h>

using namespace std;

LARGE_INTEGER timerFreq;
LARGE_INTEGER counterAtStart;

void start_time()
{
   QueryPerformanceFrequency(&timerFreq);
   QueryPerformanceCounter(&counterAtStart);
}

unsigned int get_time()
{
   if (timerFreq.QuadPart == 0)
   {
     return -1;
   }
   else
   {
      LARGE_INTEGER c;
      QueryPerformanceCounter(&c);
      return static_cast<unsigned int>( (c.QuadPart - counterAtStart.QuadPart) * 1000 / timerFreq.QuadPart );
  }
}

void swap(int* arr, int i, int j)
{
   int temp = arr[i];
   arr[i] = arr[j];
   arr[j] = temp;
}

void quick_sort(int* arr, int left, int right)
{
   int middle = (left + right) / 2;
   int pivot = arr[middle];
   int i = left;
   int j = right;

   while (i <= j)
   {
      while (arr[i] < pivot)
         i++;
      
      while (arr[j] > pivot)
         j--;

      if (i <= j)
      {
         swap(arr, i++, j--);
      }
   }
   if (left < j)
   {
      quick_sort(arr, left, j);   
   }
   if (i < right)
   {
      quick_sort(arr, i, right);   
   }
}

unsigned int sort(int len)
{
   int* arr = (int*)malloc(sizeof(int) * (len+1));
   arr[len] = NULL;

   for (int i = 0; i < len; i++)
   {
      arr[i] = (rand() % len);

      if (i % 2 == 0)
      {
         arr[i] = -arr[i];
      }
   }

   unsigned int start = get_time();
   quick_sort(arr, 0, len - 1);
   unsigned int diff = get_time() - start;

   for (int i = 1 ; i < len ; i++)
   {
      if (arr[i-1] > arr[i])
      {
         printf("not sorted!!");
         exit(1);
      }
   }

   free(arr);

   return diff;
}

int main(int argc, char **argv)
{
   timeBeginPeriod(1);
   start_time();

   srand((unsigned int)time(NULL));

   unsigned int min = 999999;
   unsigned int time;

   if (argc != 2)
   {
      printf("Usage: quicksort.exe ARRAYLENGTH");
      exit(1);
   }

   int len = atoi(argv[1]);

   for (int i = 0 ; i < 5 ; i++)
   {
      time = sort(len);
      if (time < min)
      {
         min = time;
      }
   }

   printf("c quicksort: %i (%i)", min, len);

   timeEndPeriod(1);

   return 0;
}

File: Quicksort.cs
using System;

public class Quicksort
{
   public static void Swap(int[] arr, int i, int j)
   {
      var temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
   }

   public static void Qsort(int[] arr, int left, int right)
   {
      var pivot = arr[(left + right) >> 1];
      var i = left;
      var j = right;
      while (i <= j)
      {
         while (arr[i] < pivot)
            i++;
         while (arr[j] > pivot)
            j--;
         if (i <= j)
         {
            Swap(arr, i++, j--);
         }
      }
      if (left < j)
      {
         Qsort(arr, left, j);
      }
      if (i < right)
      {
         Qsort(arr, i, right);
      }
   }

   public static double Sort(int len)
   {
      var arr = new int[len];
      var random = new Random();
      for (var i = 0; i < arr.Length; i++)
      {
         arr[i] = (int)(random.Next(len));
         if (i % 2 == 0)
         {
            arr[i] = -arr[i];
         }
      }

      var start = DateTime.Now;
      Qsort(arr, 0, arr.Length - 1);
      var diff = DateTime.Now - start;

      for (var i = 1 ; i < arr.Length ; i++)
      {
		  if (arr[i-1] > arr[i])
		  {
	         Console.WriteLine("not sorted!!");
	         Environment.Exit(1);
		  }
      }

      return diff.TotalMilliseconds;
   }

   public static void Main(string[] args)
   {
      if (args.Length != 1)
      {
         Console.WriteLine("usage: Quicksort ARRAYLENGTH");
         Environment.Exit(1);
      }

      var len = int.Parse(args[0]);

      var time = double.MaxValue;
      
      for (var i = 0 ; i < 5 ; i++)
      {
         time = Math.Min(time, Sort(len));
      }

      Console.WriteLine("C# quicksort: " + time + " ms (" + len + ")");
   }
}